lasagne/layers/special.py

```diff
@@ -544,6 +546,296 @@ def _meshgrid(height, width):
+    bilinear transformation.
+
+    Parameters
+    ----------
+    incoming : a :class:`Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape. The
+        output of this layer should be a 4D tensor, with shape
+        ``(batch_size, num_input_channels, input_rows, input_columns)``.
+
+    localization_network : a :class:`Layer` instance
+        The network that calculates the parameters of the thin plate spline
+        transformation as the x and y coordinates of the destinations of each
+        control point. The output of the localization network must be twice
+        the number of control points.
+
+    downsample_factor : float or iterable of float
+        A float or a 2-element tuple specifying the downsample factor for the
+        output image (in both spatial dimensions). A value of 1 will keep the
+        original size of the input. Values larger than 1 will downsample the
+        input. Values below 1 will upsample the input.
+
+    num_control_points : integer
+        The number of control points to be used for the thin plate spline
+        transformation. These points will be arranged as a grid along the
+        image, so the value must be a perfect square. Default is 16.
+
+    References
+    ----------
+    .. [1]  Spatial Transformer Networks
+            Max Jaderberg, Karen Simonyan, Andrew Zisserman, Koray Kavukcuoglu
+            Submitted on 5 Jun 2015
+    .. [2]  Principal warps: thin-plate splines and the decomposition of
+            deformations. Fred L. Bookstein, 1989, IEEE Transactions on
+            Pattern Analysis and Machine Intelligence.
+            http://doi.org/10.1109/34.24792
+
+    Examples
+    --------
+    Here, we'll implement an identity transform using a thin plate spline
+    transform. First we'll create the source and destination control points.
+    To make everything invariant to the shape of the image, the x and y range
+    of the image is normalized to [-1, 1] as in ref [1]_. To replicate an
+    identity transform, we'll simply set the offsets to 0. More complicated
+    transformations can easily be implemented using different x and y offsets
+    (importantly, each control point can have it's own pair of offsets).
+
+    >>> import numpy as np
+    >>> import lasagne
+    >>> num_points = 16
+    >>> grid_size = np.sqrt(num_points)
+    >>> x_source, y_source = np.meshgrid(np.linspace(-1, 1, grid_size),
+    ...                                  np.linspace(-1, 1, grid_size))
+    >>> # offset for the control points
+    >>> x_offset = 0
+    >>> y_offset = 0
+    >>> x_control_dest = x_source.flatten() + x_offset
+    >>> y_control_dest = y_source.flatten() + y_offset
+    >>> dest_points = np.vstack((x_control_dest, y_control_dest)).flatten()
+    >>>
+    >>> # We'll set the bias to be the source points
+    >>> b = np.vstack((x_source.flatten(), y_source.flatten())).flatten()
```

Comment:
> Looking through the code actually, I think it shouldn't be so hard to implement that. I have to give a presentation tomorrow, so I won't have time before then, but let me see if I can get that working tomorrow afternoon.

No need to hurry! If you think this can be done, let's just rename the argument to cater for this. It doesn't have to be part of the initial PR! Good luck with your presentation if it's not done already, otherwise I hope it went well :)